src/python/pants/engine/goal.py

```diff
@@ -0,0 +1,147 @@
+      @classproperty
+      def options_scope(cls):
+        if not outer_cls.name:
+          # TODO: Would it be unnecessarily magical to have `outer_cls.__name__.lower()` always be
+          # the name?
+          raise AssertionError('{} must have a `Goal.name` defined.'.format(outer_cls.__name__))
+        return outer_cls.name
+
+      @classmethod
+      def register_options(cls, register):
+        super(_Options, cls).register_options(register)
+        # Delegate to the outer class.
+        outer_cls.register_options(register)
+
+      @classmethod
+      def subsystem_dependencies(cls):
+        # NB: `Goal.Options` implements `SubsystemClientMixin` in order to allow v1 `Tasks` to
+        # depend on v2 Goals, and for `Goals` to declare a deprecated dependency on a `CacheSetup`
+        # instance for backwards compatibility purposes. But v2 Goals should _not_ have subsystem
+        # dependencies: instead, the @rules participating (transitively) in a Goal should directly
+        # declare their Subsystem deps.
+        if outer_cls.deprecated_cache_setup_removal_version:
+          dep = CacheSetup.scoped(
+              cls,
+              removal_version=outer_cls.deprecated_cache_setup_removal_version,
+              removal_hint='Goal `{}` uses an independent caching implementation, and ignores `{}`.'.format(
+                cls.options_scope,
+                CacheSetup.subscope(cls.options_scope),
+              )
+            )
+          return (dep,)
+        return tuple()
+
+      options_scope_category = ScopeInfo.GOAL
+
+      def __init__(self, scope, scoped_options):
+        # NB: This constructor is shaped to meet the contract of `Optionable(Factory).signature`.
+        super(_Options, self).__init__()
+        self._scope = scope
+        self._scoped_options = scoped_options
+
+      @property
+      def values(self):
+        """Returns the option values for these Goal.Options."""
+        return self._scoped_options
+    return _Options
+
+
+class _GoalOptions(object):
+  """A marker trait for the anonymous inner `Goal.Options` classes for `Goal`s."""
+
+
+class LineOriented(object):
+  """A mixin for Goal that adds Options to support the `line_oriented` context manager."""
+
+  @classmethod
+  def register_options(cls, register):
+    super(LineOriented, cls).register_options(register)
+    register('--sep', default='\\n', metavar='<separator>',
+             help='String to use to separate result lines.')
+    register('--output-file', metavar='<path>',
+             help='Write line-oriented output to this file instead.')
+
+
+@contextmanager
+def line_oriented(line_oriented_options, console):
+  """Given options and a Console, yields functions for writing to stdout and stderr, respectively.
+
+  The passed options instance will generally be the `Goal.Options` of a `LineOriented` `Goal`.
+  """
+  if not isinstance(line_oriented_options, _GoalOptions):
```

Comment:
This is a weak check. If this free function were lifted into `LineOriented` as a `@classmethod` you could check against `cls.Options` which will always be a correct check assuming `register_options` is over-ridden with proper `super` syntax. Since mixing in `LineOriented` for its option registration is already required, having this function be non-free seems like no skin off the users back and more honest.

Reply:
Yes, but I think that that would stretch the limits of `Goal.Options`. Because while it is the `Goal` that mixes-in `LineOriented`, it's the `Goal.Options` that a `@console_rule` have access to while running.

So the contextmanager would end up defined on `Goal.Options`, which would make it less discoverable. Possibly still worth it, because we would expect most consumers to start from examples?

Reply:
Like this:
```python
class LineOriented(object):
  """A mixin for Goal that adds Options to support the `line_oriented` context manager."""

   @classmethod
  def register_options(cls, register):
    super(LineOriented, cls).register_options(register)
    register('--sep', default='\\n', metavar='<separator>',
             help='String to use to separate result lines.')
    register('--output-file', metavar='<path>',
             help='Write line-oriented output to this file instead.')

  @classmethod
  @contextmanager
  def line_oriented(cls, line_oriented_options, console):
    """Given options and a Console, yields functions for writing to stdout and stderr, respectively.
    
    The passed options instance must be the `Goal.Options` of this `LineOriented` `Goal`.
    """
    if not isinstance(line_oriented_options, cls.Options):
      raise AssertionError(
          'Expected a `Goal.Options` instance for a `LineOriented` `Goal`, got: {}'.format(
            type(line_oriented_options)))

    output_file = line_oriented_options.values.output_file
    sep = line_oriented_options.values.sep.encode('utf-8').decode('unicode_escape')

    stdout, stderr = console.stdout, console.stderr
    if output_file:
      stdout = open(output_file, 'w')

    try:
      print_stdout = lambda msg: print(msg, file=stdout, end=sep)
      print_stderr = lambda msg: print(msg, file=stderr)
      yield print_stdout, print_stderr
    finally:
      if output_file:
        stdout.close()
      else:
        stdout.flush()
      stderr.flush()
```

So you'd say:
```python
class MyGoal(Goal, LineOriented):
  ...


@console_rule(MyGoal, [MyGoal.Options, Console])
def my_goal(options, console):
  with MyGoal.line_oriented(options, console) as print_stdout, print_stderr:
    ...
```

Reply:
Ahhhh. Yes. `classmethod`. Got it.

Yea, that's reasonable. Thanks!